stackrox/stackrox
# Update TypeScript 4.7.3 devDependencies in ui

## Description

* https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/
* https://github.com/microsoft/TypeScript/releases/tag/v4.7.3

### Analysis of errors

TypeScript checks some properties more carefully than before.

1. error TS2322: Type 'string | undefined' is not assignable to type 'string'.

    * src/Containers/AccessControl/AuthProviders/RuleGroups.tsx

        Indeed, `key` property is optional.

2. error TS2322: Type '{ id: string; name: string; description: string; rationale: string; remediation: string; disabled: false; categories: string[]; lifecycleStages: ("DEPLOY" | "BUILD")[]; exclusions: never[]; scope: never[]; ... 16 more ...; eventSource: "NOT_APPLICABLE"; }' is not assignable to type 'Policy'.
  Object literal may only specify known properties, and 'SORT_name' does not exist in type 'Policy'.

    * src/Containers/Policies/Modal/PolicyImport.utils.test.ts

        Indeed, 3 internal properties need to exist (or not exist) in `BasePolicy` therefore either in both or in neither of `ClientPolicy` and `Policy` also manual testing revealed incorrect camel case property names

3. Related errors
    error TS2367: This condition will always return 'false' since the types '"ACTIVE" | "INACTIVE"' and '"ATTEMPTED"' have no overlap.
    error TS2367: This condition will always return 'false' since the types '"ACTIVE" | "INACTIVE"' and '"RESOLVED"' have no overlap.

    * src/Containers/Violations/ViolationsTablePanel.tsx

        Indeed, `ListAlert` type in violationTypes.ts has incomplete string union type for `state` property.

### Changed files

1. Edit Containers/AccessControl/AuthProviders/RuleGroups.tsx
    * Replace `|| ''` conditional or with `?? ''` nullish coalescing
    * Add `?? ''` nullish coalescing

2. Rename and factor out internal properties after verifying names in responses and exported policy json file
    * Containers/Policies/Modal/PolicyImport.utils.test.ts: also delete client-only properties in `getPolicy` helper function
    * Containers/Policies/PolicyPage.tsx
    * types/policy.proto.ts

3. Edit Containers/Violations/types/violationTypes.ts
    * Import `ViolationState` from types/alert.proto.ts

### Residue

1. Replace imports from Containers/Violations/violationTypes.ts with imports from .proto.ts files
2. Edit Containers/Violations
    * Replace `LIFECYCLE_STAGES.WHATEVER` with `'WHATEVER'` because TypeScript checks both variable and constant at compile time
    * Replace `VIOLATION_STATES.WHATEVER` with `'WHATEVER'` because TypeScript checks both variable and constant at compile time
3. Delete constants/lifecycleStages.js
4. Delete constants/violationStates.js

## Checklist

## Testing Performed

1. `yarn build` in ui
2. `yarn start` in ui

### manual testing

1. RuleGroups key

    * Visit /main/access-control/auth-providers click **Create auth provider** and then select **Auth0**
    * Click **Add new rule** to see initial empty **Key** still seems okay

2. Policy internal properties

    * Visit /main/policies click a policy name link click **Actions** and then click **Export policy to JSON**
    * Go back to /main/policies click **Import policy** select the exported policy file click **Begin import** rename it and then click **Resume import**

3. Alert state

    * Visit /main/violations and click kabob menus at right of table rows

### integration testing

1. `yarn cypress-open` in ui/apps/platform
    * accessControlAuthProviders.test.js
    * policies/PatternFly/exportPolicy.test.js
    * policies/PatternFly/importPolicy.test.js
    * policies/PatternFly/policiesTable.test.js
    * policies/PatternFly/policyWizardStep3.test.js
    * violations/violations.test.js

### unit testing

1. `yarn test` in ui/apps/platform
